### PR TITLE
Add Psalm integration

### LINKED ISSUE #4: Psalm integration
### Feature Request

|    Q        |   A
|------------ | ------
| QA    | yes

#### Summary

As decided during the [Technical-Steering-Committee Meeting on August 3rd, 2020](https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-08-03-TSC-Minutes.md#vote-on-static-analysis-tooling), Laminas wants to implement [vimeo/psalm](https://github.com/vimeo/psalm) in all packages.

Implementing psalm is quite easy.

#### Required
- [x] Create a `psalm.xml` in the project root
- [x] Copy and paste the contents from this [psalm.xml.dist](https://github.com/mezzio/repo-template/blob/6124090610ef1c6d6b0c80a063cafacc3ec68f38/psalm.xml.dist)
- [x] Run `$ composer require --dev vimeo/psalm`
- [x] Run `$ vendor/bin/psalm --set-baseline=psalm-baseline.xml`
- [x] Add a composer script `static-analysis` with the command `psalm --shepherd --stats`
- [x] Add a new line to `script:` in `.travis.yml`: `- if [[ $TEST_COVERAGE == 'true' ]]; then composer static-analysis ; fi`
- [x] Remove phpstan from the project (`phpstan.neon.dist`, `.travis.yml` entry, `composer.json` `require-dev` and `scripts`)

##### Optional
- [x] Fix as many psalm errors as possible.

### PR DESCRIPTION
**Pull request by alexraputa**
Closes #4

|    Q          |   A
|-------------- | ------
| QA            | yes

### CODE REVIEW

**Review by alexraputa — COMMENTED** [review]
@boesing, what can I do with remaining Psalm issues https://travis-ci.com/github/mezzio/mezzio-authentication/jobs/396014034#L474-L499?

File: src/DefaultUser.php [hunk 1]
```diff
@@ -26,15 +26,19 @@ final class DefaultUser implements UserInterface
     private $identity;
 
     /**
-     * @var string[]
+     * @var array<mixed>
```
Comment (boesing) [inline src/DefaultUser.php]: TBH: this is less accurate than before.
If this was a list of strings before, please consider adding `@psalm-var list<string>`.
If this was a map before, consider adding `psalm-var array<string,string>`

File: src/UserRepository/PdoDatabase.php [hunk 2]
```diff
@@ -96,15 +111,16 @@ public function authenticate(string $credential, string $password = null) : ?Use
     /**
      * Get the user roles if present.
      *
-     * @param string $identity
-     * @return string[]
+     * @return array<int, mixed>
```
Comment (boesing) [inline src/UserRepository/PdoDatabase.php]: This will return a list of strings, thus, I'd prefer keeping `string[]` as it is being inferred as `array<int,string>`. 
This should not become `mixed`!

As the roles supposed to be a list of strings, please change the other typehints regarding roles to list of strings aswell.

**Review by boesing — CHANGES_REQUESTED** [review]
Nice addition here! Thank you very much.
Could you have a look on some of those `array<mixed>` annotations and re-check if we could be a bit more specific there?
At least the `roles` thing supposed to be a list of strings. 👍

File: src/UserRepository/PdoDatabase.php (hunk 2 above)
Comment (alexraputa) [reply src/UserRepository/PdoDatabase.php]: @boesing, recheck please.

File: src/DefaultUser.php (hunk 1 above)
Comment (alexraputa) [reply src/DefaultUser.php]: @boesing, recheck please.

**Review by boesing — APPROVED** [review]
LGTM.
I'd like to request some feedback from @Ocramius as I am unsure about these `empty` annotations and how they're treated in psalm.

**Comment by alexraputa** [comment]
> LGTM.
> I'd like to request some feedback from @Ocramius as I am unsure about these `empty` annotations and how they're treated in psalm.

Ok.
`TEmpty - denotes the empty type, used to describe a type corresponding to no value whatsoever. Empty arrays [] have the type array<empty, empty>.` [Psalm docs](https://psalm.dev/docs/running_psalm/plugins/plugins_type_system/#misc)

**Comment by alexraputa** [comment]
@Ocramius, friendly ping ;)

File: src/DefaultUser.php [hunk 3]
```diff
@@ -31,10 +31,14 @@ final class DefaultUser implements UserInterface
     private $roles;
 
     /**
-     * @var array
+     * @psalm-var array<string|empty, mixed>
```
Comment (Ocramius) [inline src/DefaultUser.php]: Overall, `array<string|empty, mixed>` is equivalent to `array<string, mixed>` (the empty case is included into a map case)

Let's get rid of `empty` here, since it is redundant.

Comment (Ocramius) [reply src/DefaultUser.php]: To be clear: `array<empty, empty>|array<string, mixed>` is simplified down implicitly to `array<string, mixed>` at type level

File: src/DefaultUser.php [hunk 4]
```diff
@@ -31,10 +31,14 @@ final class DefaultUser implements UserInterface
...
      */
     private $details;
 
+    /**
+     * @param string[] $roles
+     * @psalm-param array<string|empty, mixed> $details
```
Comment (Ocramius) [inline src/DefaultUser.php]: Similar to above: `empty` can be dropped

File: src/DefaultUser.php [hunk 5]
```diff
@@ -47,18 +51,24 @@ public function getIdentity() : string
...
+    /**
+     * @return string[]
+     */
     public function getRoles() : array
     {
         return $this->roles;
     }
 
+    /**
+     * @psalm-return array<string|empty, mixed>
```
Comment (Ocramius) [inline src/DefaultUser.php]: Similar to above: `empty` can be dropped

File: src/DefaultUserFactory.php [hunk 6]
```diff
@@ -22,6 +23,10 @@ class DefaultUserFactory
     public function __invoke(ContainerInterface $container) : callable
     {
         return function (string $identity, array $roles = [], array $details = []) : UserInterface {
+            Assert::isList($roles);
```
Comment (Ocramius) [inline src/DefaultUserFactory.php]: The constructor of `DefaultUser` requires a `string[]` for the `$roles`, yet here we are making it a `list<string>`.

If we don't enforce a numerically indexed array (we can't enforce it now - would be a BC break), I would remove this and avoid it.

We just limit ourselves to check that `Assert::allString($roles)`, which should suffice for our purposes

File: src/UserInterface.php [hunk 7]
```diff
@@ -20,17 +20,23 @@ public function getIdentity() : string;
     /**
      * Get all user roles
      *
-     * @return Iterable
+     * @return string[]
```
Comment (Ocramius) [inline src/UserInterface.php]: `string[]` is a legacy syntax for `array<int|string, string>`.

We return an `iterable` here, so the returned type should be `iterable<int|string, string>` to denote that we return an iterable containing strings

File: src/UserInterface.php [hunk 8]
```diff
@@ -20,17 +20,23 @@ public function getIdentity() : string;
...
+     * @param mixed|null $default
+     *
+     * @return mixed
      */
     public function getDetail(string $name, $default = null);
 
     /**
      * Get all the details, if any
+     *
+     * @psalm-return array<string|empty, mixed>
```
Comment (Ocramius) [inline src/UserInterface.php]: Similar to above: `empty` can be dropped

File: src/UserRepository/Htpasswd.php [hunk 9]
```diff
@@ -28,10 +29,18 @@ class Htpasswd implements UserRepositoryInterface
 
     /**
      * @var callable
+     *
+     * @psalm-var callable(string, array<int|empty, string|empty>, array<string|empty, mixed>): UserInterface
```
Comment (Ocramius) [inline src/UserRepository/Htpasswd.php]: Similar to above: `empty` can be dropped

File: src/UserRepository/Htpasswd.php [hunk 10]
```diff
@@ -28,10 +29,18 @@ class Htpasswd implements UserRepositoryInterface
...
      */
     private $userFactory;
 
     /**
+     * @psalm-param callable(
+     *   string,
+     *   array<int|empty, string|empty>,
```
Comment (Ocramius) [inline src/UserRepository/Htpasswd.php]: Similar to above: `empty` can be dropped

File: src/UserRepository/Htpasswd.php [hunk 11]
```diff
@@ -28,10 +29,18 @@ class Htpasswd implements UserRepositoryInterface
...
+     *   array<string|empty, mixed>
```
Comment (Ocramius) [inline src/UserRepository/Htpasswd.php]: Similar to above: `empty` can be dropped

File: src/UserRepository/Htpasswd.php [hunk 12]
```diff
@@ -45,11 +54,15 @@ public function __construct(string $filename, callable $userFactory)
         $this->filename = $filename;
 
         // Provide type safety for the composed user factory.
-        $this->userFactory = function (
+        $this->userFactory = static function (
             string $identity,
             array $roles = [],
             array $details = []
         ) use ($userFactory) : UserInterface {
+            Assert::isList($roles);
```
Comment (Ocramius) [inline src/UserRepository/Htpasswd.php]: Similar to above: we don't enforce `list<string>` on the roles, so this assertion is too restrictive

File: src/UserRepository/PdoDatabase.php [hunk 13]
```diff
@@ -29,15 +31,26 @@ class PdoDatabase implements UserRepositoryInterface
...
     /**
-     * @var array
+     * @var array<string, mixed>
      */
     private $config;
 
     /**
      * @var callable
+     *
+     * @psalm-var callable(string, array<int|empty, string|empty>, array<string|empty, mixed>): UserInterface
```
Comment (Ocramius) [inline src/UserRepository/PdoDatabase.php]: Similar to above: `empty` can be dropped

File: src/UserRepository/PdoDatabase.php [hunk 14]
```diff
@@ -29,15 +31,26 @@ class PdoDatabase implements UserRepositoryInterface
...
      */
     private $userFactory;
 
+    /**
+     * @param array<string, mixed> $config
+     *
+     * @psalm-param callable(
+     *   string,
+     *   array<int|empty, string|empty>,
+     *   array<string|empty, mixed>
```
Comment (Ocramius) [inline src/UserRepository/PdoDatabase.php]: Similar to above: `empty` can be dropped

File: src/UserRepository/PdoDatabase.php [hunk 15]
```diff
@@ -47,11 +60,15 @@ public function __construct(
         $this->config = $config;
 
         // Provide type safety for the composed user factory.
-        $this->userFactory = function (
+        $this->userFactory = static function (
             string $identity,
             array $roles = [],
             array $details = []
         ) use ($userFactory) : UserInterface {
+            Assert::isList($roles);
```
Comment (Ocramius) [inline src/UserRepository/PdoDatabase.php]: Similar to above: this restriction is too strict, as we allow `string[]` too

File: src/UserRepository/PdoDatabase.php [hunk 16]
```diff
@@ -173,6 +203,13 @@ protected function getUserDetails(string $identity) : array
         if (! $stmt->execute()) {
             return [];
         }
-        return $stmt->fetch(PDO::FETCH_ASSOC);
+
+        $userDetails = $stmt->fetch(PDO::FETCH_ASSOC);
+
+        if (! $userDetails) {
```
Comment (Ocramius) [inline src/UserRepository/PdoDatabase.php]: I think we should put in an assertion here (should be `Assert::isMap()`), as we don't expect a `false` to be returned (failure case)

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by alexraputa** [comment]
@Ocramius, thanks for review. I fixed all issues. 🎉

File: src/AuthenticationMiddleware.php [hunk 17]
```diff
@@ -18,10 +18,13 @@
 class AuthenticationMiddleware implements MiddlewareInterface
 {
     /**
-     * @var AuthenticationInterface
+     * @var AuthenticationInterface|\Zend\Expressive\Authentication\AuthenticationInterface
      */
     protected $auth;
 
+    /**
+     * @param AuthenticationInterface|\Zend\Expressive\Authentication\AuthenticationInterface $auth
```
Comment (Ocramius) [inline src/AuthenticationMiddleware.php]: Looking at this: why is it a union type, since the parameter type is enforced?

File: src/AuthenticationMiddleware.php [hunk 18]
```diff
@@ -18,10 +18,13 @@
 class AuthenticationMiddleware implements MiddlewareInterface
 {
     /**
-     * @var AuthenticationInterface
+     * @var AuthenticationInterface|\Zend\Expressive\Authentication\AuthenticationInterface
```
Comment (Ocramius) [inline src/AuthenticationMiddleware.php]: Expanding the type on a `protected` property is a BC break: is this change needed?

File: src/DefaultUser.php [hunk 19]
```diff
@@ -47,18 +51,24 @@ public function getIdentity() : string
...
+     * @psalm-return array<string, mixed>
+     */
     public function getDetails() : array
     {
         return $this->details;
     }
 
     /**
-     * @param mixed $default Default value to return if no detail matching
+     * @param mixed|null $default Default value to return if no detail matching
```
Comment (Ocramius) [inline src/DefaultUser.php]: `mixed|null` is equivalent to `mixed` (`mixed` contains the `null` type)

**Review by Ocramius — CHANGES_REQUESTED** [review]
Some more minor things that I noticed, but otherwise LGTM 👍

File: src/AuthenticationMiddleware.php (hunk 18 above)
Comment (alexraputa) [reply src/AuthenticationMiddleware.php]: I agree with you.
https://github.com/mezzio/mezzio-authentication/blob/9d3e47e4e9e5081296be2a7bc3c5dea2fb01a58c/src/AuthenticationMiddlewareFactory.php#L19-L23

But `$authentication` can be `AuthenticationInterface` or `\Zend\Expressive\Authentication\AuthenticationInterface`. 🤔

File: src/DefaultUser.php (hunk 19 above)
Comment (alexraputa) [reply src/DefaultUser.php]: Fixed.

File: src/AuthenticationMiddleware.php (hunk 18 above)
Comment (Ocramius) [reply src/AuthenticationMiddleware.php]: The `zend`-prefixed one is legacy and handled by class aliasing, so it shouldn't be required anymore

Comment (alexraputa) [reply src/AuthenticationMiddleware.php]: Can I revert this changes?

File: src/AuthenticationMiddleware.php (hunk 17 above)
Comment (alexraputa) [reply src/AuthenticationMiddleware.php]: Can I revert this changes?

Comment (Ocramius) [reply src/AuthenticationMiddleware.php]: IMO yes

File: src/AuthenticationMiddleware.php (hunk 18 above)
Comment (Ocramius) [reply src/AuthenticationMiddleware.php]: Yes

Comment (alexraputa) [reply src/AuthenticationMiddleware.php]: > Yes

Done.

File: src/AuthenticationMiddleware.php (hunk 17 above)
Comment (alexraputa) [reply src/AuthenticationMiddleware.php]: > IMO yes

Done.

File: .travis.yml [hunk 20]
```diff
@@ -46,6 +46,7 @@ install:
 script:
   - if [[ $TEST_COVERAGE == 'true' ]]; then composer test-coverage ; else composer test ; fi
   - if [[ $CS_CHECK == 'true' ]]; then composer cs-check ; fi
+  - if [[ $TEST_COVERAGE == 'true' ]]; then composer static-analysis ; fi
```
Comment (Ocramius) [inline .travis.yml:49]: Just noticed that this runs when `$TEST_COVERAGE == ' true'` - perhaps it should have its own flag? (unsure about upstream guide)

**Review by Ocramius — APPROVED** [review]
LGTM :+1: 

I think the added flag @ https://github.com/mezzio/mezzio-authentication/pull/6/files#r507527730 can be handled later - this is already tremendous value as-is!

**Comment by Ocramius** [comment]
Thanks @alexraputa! \o/

File: .travis.yml (hunk 20 above)
Comment (boesing) [reply .travis.yml:49]: Thats how it was implemented in laminas/laminas-cli#39 which was used by @weierophinney during feedback regarding the step-by-step guide. 🤷‍♂️
